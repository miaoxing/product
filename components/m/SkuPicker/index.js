import {useEffect, useState} from 'react';
import {Image, Text, View, Block} from '@fower/taro';
import {AtFloatLayout} from 'taro-ui';
import './index.scss';
import clsx from 'clsx';
import $ from 'miaoxing';
import Taro from '@tarojs/taro';
import FooterBar from '@mxjs/m-footer-bar';
import ActionButtonGroup from '../ActionButtonGroup';
import Stepper from '@mxjs/m-stepper';
import PropTypes from 'prop-types';

const contains = function (container, array) {
  for (const el of array) {
    if (!container.includes(el)) {
      return false;
    }
  }
  return true;
};

const getRange = (skus, key) => {
  if (skus.length === 1) {
    return skus[0][key];
  }

  let min = parseFloat(skus[0][key]);
  let max = min;
  skus.forEach(sku => {
    const value = parseFloat(sku[key]);
    if (value < min) {
      min = value;
    }
    if (value > max) {
      max = value;
    }
  });

  if (min === max) {
    return min;
  } else {
    return min + ' ~ ' + max;
  }
};

const getPrice = function (validSkus) {
  let price = getRange(validSkus, 'price');
  let score = getRange(validSkus, 'score');

  return (
    <Block>
      {!!price && <Text textXs>￥</Text>}
      {!!price && price}
      {price && score ? ' + ' : ''}
      {!!score && score + '积分'}
    </Block>
  );
};

// TODO 简化
const calSoldOutValueIds = function (skus, ids = []) {
  // 1. 找出所有售罄的规格值编号
  const valueStockNums = {};
  skus.forEach(sku => {
    sku.specValueIds.forEach(valueId => {
      if (!valueStockNums[valueId]) {
        valueStockNums[valueId] = 0;
      }
      valueStockNums[valueId] += sku.stockNum;
    });
  });

  let soldOutValueIds = [];
  for (const valueId in valueStockNums) {
    if (valueStockNums[valueId] === 0) {
      soldOutValueIds.push(parseInt(valueId, 10));
    }
  }

  // 2. 找出选中后，剩下的是售罄的规则值编号
  if (ids.length) {
    skus.forEach(sku => {
      if (sku.stockNum !== 0) {
        return;
      }
      ids.forEach(id => {
        if (sku.specValueIds.includes(id)) {
          sku.specValueIds.forEach(valueId => {
            if (valueId === id) {
              return;
            }
            soldOutValueIds.push(valueId);
          });
        }
      });
    });
  }

  return soldOutValueIds;
};

const generateSelectedText = (product, selectedValueIds) => {
  const requiredSpecNames = [];
  const selectedValueNames = [];
  product.spec.specs.map(spec => {
    if (!selectedValueIds[spec.id]) {
      requiredSpecNames.push(spec.name);
      return;
    }
    spec.values.forEach(value => {
      if (value.id === selectedValueIds[spec.id]) {
        selectedValueNames.push(value.name);
      }
    });
  });

  // 还有规格没选完，提示没选完的规格名称
  if (requiredSpecNames.length) {
    return requiredSpecNames.join(' / ');
  }

  // 全部规格选完，提示已选的规格值
  return '已选：' + selectedValueNames.join(' / ');
};

/**
 * 通过规格值数组，计算出规格值对象
 */
const calSelectedValueIds = (product, ids) => {
  const selectedValueIds = {};

  product.spec.specs.forEach(spec => {
    spec.values.forEach(value => {
      if (ids.includes(value.id)) {
        selectedValueIds[spec.id] = value.id;
      }
    });
  });
  return selectedValueIds;
};

const useStateWithDep = (defaultValue, dep) => {
  const hasDep = typeof dep !== 'undefined';
  const [value, setValue] = useState(defaultValue);

  useEffect(() => {
    setValue(hasDep ? defaultValue() : defaultValue);
  }, [JSON.stringify(hasDep ? dep : defaultValue)]);
  return [value, setValue];
};

const SkuPicker = (
  {
    product,
    isOpened,
    onClose,
    action,
    quantity: quantityProp = 1,
    selectedValueIds: selectedValueIdsProp = [],
    cartId,
    onAfterSelectSpec,
    onAfterRequest,
  },
) => {
  // 当前可购买的数量
  const [stockNum, setStockNum] = useState();

  // 当前的价格范围
  const [price, setPrice] = useState();

  // 购买的数量
  const [quantity, setQuantity] = useStateWithDep(quantityProp);

  // 当前选中的规格值编号
  const [selectedValueIds, setSelectedValueIds] = useStateWithDep(() => {
    return calSelectedValueIds(product, selectedValueIdsProp);
  }, selectedValueIdsProp);

  // 选中所有规格后计算出的 SKU 编号
  const [skuId, setSkuId] = useState();

  // 售罄的规格值编号
  const [soldOutValueIds, setSoldOutValueIds] = useState(() => calSoldOutValueIds(product.skus));

  // 选择/取消选择规格值
  const selectValueId = (i, valueId) => {
    const isUnselect = selectedValueIds[i] === valueId;

    // 允许取消选择，但不可选中已售罄的规格值
    if (!isUnselect && soldOutValueIds.includes(valueId)) {
      return;
    }

    if (isUnselect) {
      delete selectedValueIds[i];
    } else {
      selectedValueIds[i] = valueId;
    }
    setSelectedValueIds({...selectedValueIds});
  };

  const handleClose = () => {
    onClose && onClose();
  };

  // 预设接口允许外部调用
  const api = {};

  api.generateSelectedText = () => {
    return generateSelectedText(product, selectedValueIds);
  };

  api.createCart = async () => {
    const {ret} = await $.post({
      url: $.apiUrl('carts'),
      data: {
        skuId,
        quantity,
      },
    });

    $.ret(ret);
    ret.isSuc() && handleClose();
    onAfterRequest && onAfterRequest(ret);
  };

  api.updateCart = async () => {
    const {ret} = await $.patch({
      url: $.apiUrl('carts/%s', cartId),
      data: {
        skuId,
        quantity,
      },
    });

    $.ret(ret);
    ret.isSuc() && handleClose();
    onAfterRequest && onAfterRequest(ret);
  };

  api.createOrder = () => {
    handleClose();
    Taro.navigateTo({
      url: $.url('orders/new', {skuId, quantity}),
    });
  };

  // 选择规格后，更新 SKU 编号, 库存数量，价格文案等
  useEffect(() => {
    const ids = Object.values(selectedValueIds);

    let stockNum = 0;
    let validSkus = [];
    if (ids.length === 0) {
      stockNum = product.stockNum;
      validSkus = product.skus;
    } else {
      product.skus.forEach(sku => {
        if (contains(sku.specValueIds, ids)) {
          stockNum += sku.stockNum;
          validSkus.push(sku);
        }
      });
    }

    setSkuId(validSkus.length === 1 ? validSkus[0].id : null);
    setSoldOutValueIds(calSoldOutValueIds(product.skus, ids));
    setStockNum(stockNum);
    setQuantity(stockNum < quantity ? stockNum : quantity);
    setPrice(getPrice(validSkus));
    onAfterSelectSpec && onAfterSelectSpec(api);
  }, [JSON.stringify(selectedValueIds)]);

  // 检查规格是否已选完
  const check = () => {
    for (const spec of product.spec.specs) {
      if (!selectedValueIds[spec.id]) {
        $.err('请选择' + spec.name);
        return false;
      }
    }
    return true;
  };

  const handleClickButton = (action) => {
    if (!check()) {
      return;
    }
    if (action === 'createCart') {
      api.createCart();
    } else if (action === 'updateCart') {
      api.updateCart();
    } else {
      api.createOrder();
    }
  };

  const handlePreviewImage = () => {
    Taro.previewImage({
      urls: [product.image],
    });
  };

  const unit = product.configs.unit || '件';

  return (
    <AtFloatLayout isOpened={isOpened} onClose={handleClose} className="sku-picker">
      <View m3 toBetween>
        <View flex>
          <Image w16 h16 rounded1 overflowHidden src={product.image} onClick={handlePreviewImage}/>
          <View mx2 textBase brand500 alignSelf="flex-end">
            {price}
          </View>
        </View>
        <View text3XL fontHairline mt="-1rem" gray500 onClick={handleClose}>&times;</View>
      </View>

      <View maxH="60vh" overflowYScroll mb="56px">
        {!product.spec.isDefault && <View m3 column className="border-b">
          {product.spec.specs.map((spec) => {
            return (
              <View key={spec.id} className="sku-picker-spec-item">
                <View className="sku-picker-spec-name">{spec.name}</View>
                <View>
                  {spec.values.map(value => {
                    const active = selectedValueIds[spec.id] === value.id;
                    return (
                      <Text
                        key={value.id}
                        className={clsx('sku-picker-spec-value', {
                          active,
                          // 购物车中，售罄时，不禁用已选中的规格
                          disabled: !active && soldOutValueIds.includes(value.id),
                        })}
                        onClick={selectValueId.bind(this, spec.id, value.id)}
                      >
                        {value.name}
                      </Text>
                    );
                  })}
                </View>
              </View>
            );
          })}
        </View>}

        <View m3 toCenterY toBetween>
          <Text mr2 textSM css={{whiteSpace: 'nowrap'}}>数量</Text>
          <View toCenterY toBetween>
            <Text gray500 textXS mr2>
              {product.maxOrderQuantity > 0 && `每人限购 ${product.maxOrderQuantity} ${unit}`}
              {' '}
              剩下 {stockNum} {unit}
            </Text>
            {' '}
            <Stepper value={quantity} onChange={setQuantity} min={1} max={stockNum}/>
          </View>
        </View>
      </View>

      <FooterBar>
        <ActionButtonGroup ret={product.createCartOrOrder} action={action} onClick={handleClickButton}/>
      </FooterBar>
    </AtFloatLayout>
  );
};

SkuPicker.propTypes = {
  product: PropTypes.object.isRequired,
  isOpened: PropTypes.bool,
  onClose: PropTypes.func,
  action: PropTypes.string,
  quantity: PropTypes.number,
  selectedValueIds: PropTypes.arrayOf(PropTypes.number),
  cartId: PropTypes.number,
  onAfterSelectSpec: PropTypes.func,
  onAfterRequest: PropTypes.func,
};

export default SkuPicker;
