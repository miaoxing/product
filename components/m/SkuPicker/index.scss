@import "~taro-ui/dist/style/components/float-layout.scss";

.at-float-layout {
  &__container,
  & .layout-body,
  & .layout-body__content {
    // TODO 改为固定高度，以便每次打开都在同一位置
    // 加大窗口最大高度
    max-height: 99vh;
    // 不限窗口最小高度
    min-height: 0;
  }
}

.sku-picker {
  .at-float-layout__container {
    // 头部使用圆角更加美观
    border-top-left-radius: 1rem;
    border-top-right-radius: 1rem;
  }

  // 内部元素自己控制间距
  .layout-body {
    padding: 0;
  }

  &-spec-item {
    margin-bottom: .75rem;
    overflow: hidden;
  }

  &-spec-name {
    margin-bottom: .25rem;
    font-size: 0.875rem;
  }

  &-spec-value {
    float: left;
    border-radius: .2rem;
    text-align: center;
    font-size: .75rem;
    line-height: 1;
    min-width: 1.5rem;
    padding: .5rem 0.75rem;
    margin-bottom: .5rem;
    margin-right: .5rem;
    border: 1PX solid var(--mx-colors-gray-100);
    background: var(--mx-colors-gray-100);

    &.active {
      color: var(--mx-colors-primary-500);
      border-color: var(--mx-colors-primary-500);
      background: var(--mx-colors-primary-50);
    }

    &.disabled {
      color: var(--mx-colors-gray-400);
    }
  }
}
