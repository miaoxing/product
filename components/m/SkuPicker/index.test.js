import SkuPicker from './index';
import {render, fireEvent} from '@testing-library/react';
import {Ret} from 'miaoxing';
import {useState} from 'react';

const product = {
  spec: {
    specs: [
      {
        id: 3,
        name: '默认',
        values: [
          {
            id: 4,
            name: '默认',
          },
        ],
      },
    ],
  },
  skus: [
    {
      id: 1,
      specValueIds: [4],
      stockNum: 10,
    },
  ],
  configs: {
    unit: '件',
  },
  createCartOrOrder: Ret.suc({
    createCart: Ret.suc(),
    createOrder: Ret.suc(),
  }),
};

describe('SkuPicker', () => {
  test('basic', async () => {
    const {container} = render(<SkuPicker product={product}/>);

    expect(container).toMatchSnapshot();
  });

  test('close', () => {
    const handleClose = jest.fn();
    const {getByText} = render(<SkuPicker isOpened={true} onClose={handleClose} product={product}/>);

    const close = getByText('×');
    fireEvent.click(close);

    expect(handleClose).toBeCalledTimes(1);
    expect(handleClose.mock.calls).toMatchSnapshot();
  });
});
