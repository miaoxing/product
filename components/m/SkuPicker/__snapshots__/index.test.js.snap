// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkuPicker basic 1`] = `
<div>
  <taro-view-core
    class="at-float-layout sku-picker"
  >
    <taro-view-core
      class="at-float-layout__overlay"
    />
    <taro-view-core
      class="at-float-layout__container layout"
    >
      <taro-view-core
        class="layout-body"
      >
        <taro-scroll-view-core
          class="layout-body__content"
          scrolly="true"
        >
          <taro-view-core
            style="justify-content: space-between; flex-direction: row; display: flex;"
          >
            <taro-view-core
              style="display: flex;"
            >
              <taro-image-core
                style="border-radius: 4rpx; overflow: hidden;"
              />
              <taro-view-core
                style="color: rgb(32, 201, 151); align-self: flex-end;"
              >
                <taro-block-core>
                  
                </taro-block-core>
              </taro-view-core>
            </taro-view-core>
            <taro-view-core
              style="margin-top: -1rem; font-weight: 100; color: rgb(113, 113, 122);"
            >
              ×
            </taro-view-core>
          </taro-view-core>
          <taro-view-core
            style="margin-bottom: 56px; max-height: 60vh; overflow-y: scroll;"
          >
            <taro-view-core
              class="border-b"
              style="flex-direction: column; display: flex;"
            >
              <taro-view-core
                class="sku-picker-spec-item"
              >
                <taro-view-core
                  class="sku-picker-spec-name"
                >
                  默认
                </taro-view-core>
                <taro-view-core>
                  <taro-text-core
                    class="sku-picker-spec-value active"
                  >
                    默认
                  </taro-text-core>
                </taro-view-core>
              </taro-view-core>
            </taro-view-core>
            <taro-view-core
              style="align-items: center; justify-content: space-between; flex-direction: row; display: flex;"
            >
              <taro-text-core
                class="css-1vq2xix-SkuPicker"
              >
                数量
              </taro-text-core>
              <taro-view-core
                style="align-items: center; justify-content: space-between; flex-direction: row; display: flex;"
              >
                <taro-text-core
                  style="color: rgb(113, 113, 122);"
                >
                   
                  剩下 
                  10
                   
                  件
                </taro-text-core>
                 
                <taro-view-core
                  class="mx-stepper"
                >
                  <taro-button-core
                    class="mx-stepper-minus"
                    disabled="true"
                  />
                  <taro-input-core
                    class="mx-stepper-input"
                  />
                  <taro-button-core
                    class="mx-stepper-plus"
                  />
                </taro-view-core>
              </taro-view-core>
            </taro-view-core>
          </taro-view-core>
          <taro-view-core
            class="mx-footer-bar"
          >
            <taro-view-core
              class="mx-btn-theme mx-btn-theme-pill"
            >
              <taro-view-core
                class="mx-btn-group"
              >
                <taro-button-core
                  class="mx-btn mx-btn-secondary mx-footer-bar-btn"
                >
                  加入购物车
                </taro-button-core>
                <taro-button-core
                  class="mx-btn mx-btn-primary mx-footer-bar-btn"
                >
                  立即购买
                </taro-button-core>
              </taro-view-core>
            </taro-view-core>
          </taro-view-core>
        </taro-scroll-view-core>
      </taro-view-core>
    </taro-view-core>
  </taro-view-core>
</div>
`;

exports[`SkuPicker close 1`] = `
Array [
  Array [],
]
`;
